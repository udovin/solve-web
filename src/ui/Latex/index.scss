@import "katex/dist/katex.min.css";
@import "../../mixins.scss";

.ui-latex {
	.lstlisting {
		padding: 10px;
		border-radius: 4px;
		font-family: monospace;
		overflow: auto;
	}

	.texttt {
		font-family: monospace;
	}

	img {
		max-width: 100%;
	}
}

@include themify {
	.ui-latex {
		.lstlisting {
			color: themed("button-active-text-color");
			background-color: themed("button-active-background-color");
		}
	}
}
