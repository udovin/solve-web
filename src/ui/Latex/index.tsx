import { FC, useCallback } from "react";
import { Macro } from "@unified-latex/unified-latex-types";
import { unifiedLatexFromString } from "@unified-latex/unified-latex-util-parse";
import { convertToHtml } from "@unified-latex/unified-latex-to-hast";
import { htmlLike } from "@unified-latex/unified-latex-util-html-like";
import { s } from "@unified-latex/unified-latex-builder";
import { replaceNode } from "@unified-latex/unified-latex-util-replace";
import { printRaw } from "@unified-latex/unified-latex-util-print-raw";
import { pgfkeysArgToObject } from "@unified-latex/unified-latex-util-pgfkeys";
import { VisitInfo } from "@unified-latex/unified-latex-util-visit";
import { unified } from "unified";
import katex from "katex";

import "./index.scss";

export type LatexProps = {
	className?: string;
	content?: string;
	imageBaseUrl?: string;
};

const parser = unified().use(unifiedLatexFromString, {
	macros: {
		"def": { signature: "m m" },
		"underline": { signature: "m", renderInfo: { inParMode: true } },
	},
});

type Context = {
	imageBaseUrl?: string;
};

const macros: Record<string, (node: Macro, info: VisitInfo, context: Context) => any> = {
	"def": () => null,
	"includegraphics": (node: Macro, _: VisitInfo, context: Context) => {
		if (!node.args) {
			return null;
		}
		const args = pgfkeysArgToObject(node.args[1]);
		let style = '';
		if (args["width"] && args["width"].length) {
			style += `width:${printRaw(args["width"][0])};`;
		}
		if (args["height"] && args["height"].length) {
			style += `height:${printRaw(args["height"][0])};`;
		}
		let scale: string | undefined = undefined;
		if (args["scale"] && args["scale"].length) {
			style += `display:none;`
			scale = printRaw(args["scale"][0]);
		}
		const imageName = printRaw(node.args[3].content);
		const imageUrl = (context.imageBaseUrl ?? "") + imageName;
		let attributes: object = { src: imageUrl, style: style };
		if (scale) {
			attributes = { ...attributes, style: style, "data-scale": scale };
		}
		return htmlLike({ tag: "img", attributes });
	},
	"underline": (node: Macro) => {
		if (!node.args) {
			return null;
		}
		const args = node.args.map((arg) => {
			if (arg.openMark === "" && arg.content.length === 0) {
				return null;
			}
			return arg.content;
		});
		const content = args[args.length - 1] || [];
		const attributes = { className: "underline" };
		return htmlLike({ tag: "u", content, attributes });
	},
	"^": (_: Macro, info: VisitInfo) => {
		if (info.context.inMathMode) {
			return undefined;
		}
		if (info.index === undefined || !info.containingArray) {
			return undefined;
		}
		if (info.containingArray.length <= info.index + 1) {
			return undefined;
		}
		const nextToken = info.containingArray[info.index + 1];
		const content = printRaw(nextToken);
		if (content !== "{}") {
			return undefined;
		}
		return s("^");
	},
	"bf": (_: Macro, info: VisitInfo) => {
		if (!info.context.hasMathModeAncestor) {
			return undefined;
		}
		return s("\\bf");
	},
};

const Latex: FC<LatexProps> = props => {
	const { className, content, imageBaseUrl } = props;
	const ast = parser.parse(content ?? "");
	const context = {
		imageBaseUrl: imageBaseUrl,
	};
	replaceNode(ast, (node, info) => {
		if (node.type !== "macro") {
			return undefined;
		}
		const macro = macros[node.content];
		if (!macro) {
			return undefined;
		}
		return macro(node, info, context);
	});
	const html = convertToHtml(ast);
	const ref = useCallback((node: HTMLDivElement) => {
		if (!node) {
			return;
		}
		node.querySelectorAll(".display-math").forEach(dm => {
			katex.render(dm.textContent ?? "", dm as HTMLElement, {
				displayMode: true,
				throwOnError: false,
			});
		});
		node.querySelectorAll(".inline-math").forEach(im => {
			katex.render(im.textContent ?? "", im as HTMLElement, {
				displayMode: false,
				throwOnError: false,
			});
		});
		node.querySelectorAll<HTMLImageElement>("img[data-scale]").forEach(img => {
			const scaleAttr = img.getAttribute("data-scale");
			if (scaleAttr) {
				const scale = parseFloat(scaleAttr);
				img.onload = () => {
					img.style.setProperty("width", `${img.naturalWidth * scale}px`);
					img.style.removeProperty("display");
				};
			}
		});
	}, []);
	return <div
		className={`ui-latex ${className ?? ""}`}
		ref={ref}
		dangerouslySetInnerHTML={{ __html: html }}
	/>;
};

export default Latex;
