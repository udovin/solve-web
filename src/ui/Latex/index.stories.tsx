import { FC, useState } from "react";
import Latex, { LatexProps } from ".";

export default {
	title: "Latex",
};

export const Index = () => <>
	<Latex content={`\\section{Fermat's Last Theorem} $a^n + b^n \\ne c^n$, $a, b, c, n \\in \\N$, $n \\ge 3$`} />
	<Latex content={`a\\^{}b = $a^b$`} />
	<Latex content={`\\^a, \\^{a}, \\^{}a`} />
</>;
