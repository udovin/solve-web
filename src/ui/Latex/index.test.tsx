import { render } from "@testing-library/react";

import Latex from ".";

test("renders correctly", () => {
	render(<Latex content={`\\section{Fermat's Last Theorem} $a^n + b^n \\ne c^n$, $a, b, c, n \\in \\N$, $n \\ge 3$`} />);
	expect(document.body).toMatchSnapshot();
});

test("renders correctly", () => {
	render(<Latex content={`a\\^{}b = $a^b$`} />);
	expect(document.body).toMatchSnapshot();
});

test("renders correctly", () => {
	render(<Latex content={`\\^a, \\^{a}, \\^{}a`} />);
	expect(document.body).toMatchSnapshot();
});

test("renders correctly", () => {
	render(<Latex content={`\\underline{Underline} {\\bfseries Bold} $\\bf{Bold2}$ {\\bf Bold3} Normal`} />);
	expect(document.body).toMatchSnapshot();
});
