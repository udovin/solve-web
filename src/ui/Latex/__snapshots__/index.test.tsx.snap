// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<body>
  <div>
    <div
      class="ui-latex "
    >
      <h3>
        Fermat’s Last Theorem
      </h3>
       
      <span
        class="inline-math"
      >
        <span
          class="katex"
        >
          <span
            class="katex-mathml"
          >
            <math
              xmlns="http://www.w3.org/1998/Math/MathML"
            >
              <semantics>
                <mrow>
                  <msup>
                    <mi>
                      a
                    </mi>
                    <mi>
                      n
                    </mi>
                  </msup>
                  <mo>
                    +
                  </mo>
                  <msup>
                    <mi>
                      b
                    </mi>
                    <mi>
                      n
                    </mi>
                  </msup>
                  <mo
                    mathvariant="normal"
                  >
                    ≠
                  </mo>
                  <msup>
                    <mi>
                      c
                    </mi>
                    <mi>
                      n
                    </mi>
                  </msup>
                </mrow>
                <annotation
                  encoding="application/x-tex"
                >
                  a^{n} + b^{n} \\ne c^{n}
                </annotation>
              </semantics>
            </math>
          </span>
          <span
            aria-hidden="true"
            class="katex-html"
          >
            <span
              class="base"
            >
              <span
                class="strut"
                style="height: 0.7477em; vertical-align: -0.0833em;"
              />
              <span
                class="mord"
              >
                <span
                  class="mord mathnormal"
                >
                  a
                </span>
                <span
                  class="msupsub"
                >
                  <span
                    class="vlist-t"
                  >
                    <span
                      class="vlist-r"
                    >
                      <span
                        class="vlist"
                        style="height: 0.6644em;"
                      >
                        <span
                          class=""
                          style="top: -3.063em; margin-right: 0.05em;"
                        >
                          <span
                            class="pstrut"
                            style="height: 2.7em;"
                          />
                          <span
                            class="sizing reset-size6 size3 mtight"
                          >
                            <span
                              class="mord mtight"
                            >
                              <span
                                class="mord mathnormal mtight"
                              >
                                n
                              </span>
                            </span>
                          </span>
                        </span>
                      </span>
                    </span>
                  </span>
                </span>
              </span>
              <span
                class="mspace"
                style="margin-right: 0.2222em;"
              />
              <span
                class="mbin"
              >
                +
              </span>
              <span
                class="mspace"
                style="margin-right: 0.2222em;"
              />
            </span>
            <span
              class="base"
            >
              <span
                class="strut"
                style="height: 0.8889em; vertical-align: -0.1944em;"
              />
              <span
                class="mord"
              >
                <span
                  class="mord mathnormal"
                >
                  b
                </span>
                <span
                  class="msupsub"
                >
                  <span
                    class="vlist-t"
                  >
                    <span
                      class="vlist-r"
                    >
                      <span
                        class="vlist"
                        style="height: 0.6644em;"
                      >
                        <span
                          class=""
                          style="top: -3.063em; margin-right: 0.05em;"
                        >
                          <span
                            class="pstrut"
                            style="height: 2.7em;"
                          />
                          <span
                            class="sizing reset-size6 size3 mtight"
                          >
                            <span
                              class="mord mtight"
                            >
                              <span
                                class="mord mathnormal mtight"
                              >
                                n
                              </span>
                            </span>
                          </span>
                        </span>
                      </span>
                    </span>
                  </span>
                </span>
              </span>
              <span
                class="mspace"
                style="margin-right: 0.2778em;"
              />
              <span
                class="mrel"
              >
                <span
                  class="mrel"
                >
                  <span
                    class="mord vbox"
                  >
                    <span
                      class="thinbox"
                    >
                      <span
                        class="rlap"
                      >
                        <span
                          class="strut"
                          style="height: 0.8889em; vertical-align: -0.1944em;"
                        />
                        <span
                          class="inner"
                        >
                          <span
                            class="mord"
                          >
                            <span
                              class="mrel"
                            >
                              
                            </span>
                          </span>
                        </span>
                        <span
                          class="fix"
                        />
                      </span>
                    </span>
                  </span>
                </span>
                <span
                  class="mrel"
                >
                  =
                </span>
              </span>
              <span
                class="mspace"
                style="margin-right: 0.2778em;"
              />
            </span>
            <span
              class="base"
            >
              <span
                class="strut"
                style="height: 0.6644em;"
              />
              <span
                class="mord"
              >
                <span
                  class="mord mathnormal"
                >
                  c
                </span>
                <span
                  class="msupsub"
                >
                  <span
                    class="vlist-t"
                  >
                    <span
                      class="vlist-r"
                    >
                      <span
                        class="vlist"
                        style="height: 0.6644em;"
                      >
                        <span
                          class=""
                          style="top: -3.063em; margin-right: 0.05em;"
                        >
                          <span
                            class="pstrut"
                            style="height: 2.7em;"
                          />
                          <span
                            class="sizing reset-size6 size3 mtight"
                          >
                            <span
                              class="mord mtight"
                            >
                              <span
                                class="mord mathnormal mtight"
                              >
                                n
                              </span>
                            </span>
                          </span>
                        </span>
                      </span>
                    </span>
                  </span>
                </span>
              </span>
            </span>
          </span>
        </span>
      </span>
      , 
      <span
        class="inline-math"
      >
        <span
          class="katex"
        >
          <span
            class="katex-mathml"
          >
            <math
              xmlns="http://www.w3.org/1998/Math/MathML"
            >
              <semantics>
                <mrow>
                  <mi>
                    a
                  </mi>
                  <mo
                    separator="true"
                  >
                    ,
                  </mo>
                  <mi>
                    b
                  </mi>
                  <mo
                    separator="true"
                  >
                    ,
                  </mo>
                  <mi>
                    c
                  </mi>
                  <mo
                    separator="true"
                  >
                    ,
                  </mo>
                  <mi>
                    n
                  </mi>
                  <mo>
                    ∈
                  </mo>
                  <mi
                    mathvariant="double-struck"
                  >
                    N
                  </mi>
                </mrow>
                <annotation
                  encoding="application/x-tex"
                >
                  a, b, c, n \\in \\N
                </annotation>
              </semantics>
            </math>
          </span>
          <span
            aria-hidden="true"
            class="katex-html"
          >
            <span
              class="base"
            >
              <span
                class="strut"
                style="height: 0.8889em; vertical-align: -0.1944em;"
              />
              <span
                class="mord mathnormal"
              >
                a
              </span>
              <span
                class="mpunct"
              >
                ,
              </span>
              <span
                class="mspace"
                style="margin-right: 0.1667em;"
              />
              <span
                class="mord mathnormal"
              >
                b
              </span>
              <span
                class="mpunct"
              >
                ,
              </span>
              <span
                class="mspace"
                style="margin-right: 0.1667em;"
              />
              <span
                class="mord mathnormal"
              >
                c
              </span>
              <span
                class="mpunct"
              >
                ,
              </span>
              <span
                class="mspace"
                style="margin-right: 0.1667em;"
              />
              <span
                class="mord mathnormal"
              >
                n
              </span>
              <span
                class="mspace"
                style="margin-right: 0.2778em;"
              />
              <span
                class="mrel"
              >
                ∈
              </span>
              <span
                class="mspace"
                style="margin-right: 0.2778em;"
              />
            </span>
            <span
              class="base"
            >
              <span
                class="strut"
                style="height: 0.6889em;"
              />
              <span
                class="mord mathbb"
              >
                N
              </span>
            </span>
          </span>
        </span>
      </span>
      , 
      <span
        class="inline-math"
      >
        <span
          class="katex"
        >
          <span
            class="katex-mathml"
          >
            <math
              xmlns="http://www.w3.org/1998/Math/MathML"
            >
              <semantics>
                <mrow>
                  <mi>
                    n
                  </mi>
                  <mo>
                    ≥
                  </mo>
                  <mn>
                    3
                  </mn>
                </mrow>
                <annotation
                  encoding="application/x-tex"
                >
                  n \\ge 3
                </annotation>
              </semantics>
            </math>
          </span>
          <span
            aria-hidden="true"
            class="katex-html"
          >
            <span
              class="base"
            >
              <span
                class="strut"
                style="height: 0.7719em; vertical-align: -0.136em;"
              />
              <span
                class="mord mathnormal"
              >
                n
              </span>
              <span
                class="mspace"
                style="margin-right: 0.2778em;"
              />
              <span
                class="mrel"
              >
                ≥
              </span>
              <span
                class="mspace"
                style="margin-right: 0.2778em;"
              />
            </span>
            <span
              class="base"
            >
              <span
                class="strut"
                style="height: 0.6444em;"
              />
              <span
                class="mord"
              >
                3
              </span>
            </span>
          </span>
        </span>
      </span>
    </div>
  </div>
</body>
`;

exports[`renders correctly 2`] = `
<body>
  <div>
    <div
      class="ui-latex "
    >
      a^b = 
      <span
        class="inline-math"
      >
        <span
          class="katex"
        >
          <span
            class="katex-mathml"
          >
            <math
              xmlns="http://www.w3.org/1998/Math/MathML"
            >
              <semantics>
                <mrow>
                  <msup>
                    <mi>
                      a
                    </mi>
                    <mi>
                      b
                    </mi>
                  </msup>
                </mrow>
                <annotation
                  encoding="application/x-tex"
                >
                  a^{b}
                </annotation>
              </semantics>
            </math>
          </span>
          <span
            aria-hidden="true"
            class="katex-html"
          >
            <span
              class="base"
            >
              <span
                class="strut"
                style="height: 0.8491em;"
              />
              <span
                class="mord"
              >
                <span
                  class="mord mathnormal"
                >
                  a
                </span>
                <span
                  class="msupsub"
                >
                  <span
                    class="vlist-t"
                  >
                    <span
                      class="vlist-r"
                    >
                      <span
                        class="vlist"
                        style="height: 0.8491em;"
                      >
                        <span
                          class=""
                          style="top: -3.063em; margin-right: 0.05em;"
                        >
                          <span
                            class="pstrut"
                            style="height: 2.7em;"
                          />
                          <span
                            class="sizing reset-size6 size3 mtight"
                          >
                            <span
                              class="mord mtight"
                            >
                              <span
                                class="mord mathnormal mtight"
                              >
                                b
                              </span>
                            </span>
                          </span>
                        </span>
                      </span>
                    </span>
                  </span>
                </span>
              </span>
            </span>
          </span>
        </span>
      </span>
    </div>
  </div>
</body>
`;

exports[`renders correctly 3`] = `
<body>
  <div>
    <div
      class="ui-latex "
    >
      â, â, ^a
    </div>
  </div>
</body>
`;

exports[`renders correctly 4`] = `
<body>
  <div>
    <div
      class="ui-latex "
    >
      <u
        class="underline"
      >
        Underline
      </u>
       
      <b
        class="textbf"
      >
        Bold
      </b>
       
      <span
        class="inline-math"
      >
        <span
          class="katex"
        >
          <span
            class="katex-mathml"
          >
            <math
              xmlns="http://www.w3.org/1998/Math/MathML"
            >
              <semantics>
                <mrow>
                  <mi
                    mathvariant="bold"
                  >
                    B
                  </mi>
                  <mi
                    mathvariant="bold"
                  >
                    o
                  </mi>
                  <mi
                    mathvariant="bold"
                  >
                    l
                  </mi>
                  <mi
                    mathvariant="bold"
                  >
                    d
                  </mi>
                  <mn
                    mathvariant="bold"
                  >
                    2
                  </mn>
                </mrow>
                <annotation
                  encoding="application/x-tex"
                >
                  \\bf{Bold2}
                </annotation>
              </semantics>
            </math>
          </span>
          <span
            aria-hidden="true"
            class="katex-html"
          >
            <span
              class="base"
            >
              <span
                class="strut"
                style="height: 0.6944em;"
              />
              <span
                class="mord"
              >
                <span
                  class="mord"
                >
                  <span
                    class="mord mathbf"
                  >
                    Bold2
                  </span>
                </span>
              </span>
            </span>
          </span>
        </span>
      </span>
       
      <b
        class="textbf"
      >
        Bold3
      </b>
       Normal
    </div>
  </div>
</body>
`;
